copycat/api/src/main/java/org/apache/kafka/copycat/data/SchemaProjector.java

```diff
@@ -0,0 +1,209 @@
+            }
+        } else {
+            if (record != null) {
+                return projectRequiredSchema(writer, reader, record);
+            }
+            return null;
+        }
+    }
+
+    private static Object projectRequiredSchema(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        Object result = null;
+        // handle logical types
+        if (reader.name() != null) {
+            return projectLogical(writer, reader, record);
+        }
+        switch (reader.type()) {
+            case INT8:
+            case INT16:
+            case INT32:
+            case INT64:
+            case FLOAT32:
+            case FLOAT64:
+            case BOOLEAN:
+            case BYTES:
+            case STRING:
+                result = projectPrimitive(writer, reader, record);
+                break;
+            case STRUCT:
+                result = projectStruct(writer, reader, (Struct) record);
+                break;
+            case ARRAY:
+                result = projectArray(writer, reader, record);
+                break;
+            case MAP:
+                result = projectMap(writer, reader, record);
+        }
+        return result;
+    }
+
+    private static Object projectStruct(Schema writer, Schema reader, Struct writerStruct) throws SchemaProjectorException {
+        Struct readerStruct = new Struct(reader);
+        for (Field readerField : reader.fields()) {
+            String fieldName = readerField.name();
+            Field writerField = writer.field(fieldName);
+            if (writerField != null) {
+                Object writerFieldValue = writerStruct.get(fieldName);
+                Object readerFieldValue = project(writerField.schema(), readerField.schema(), writerFieldValue);
+                readerStruct.put(fieldName, readerFieldValue);
+            } else {
+                Object readerDefault;
+                if (readerField.schema().defaultValue() != null) {
+                    readerDefault = readerField.schema().defaultValue();
+                } else {
+                    throw new SchemaProjectorException("Cannot project " + writer.schema() + " to " + reader.schema());
+                }
+                readerStruct.put(fieldName, readerDefault);
+            }
+        }
+        return readerStruct;
+    }
+
+    private static Object projectArray(Schema writer, Schema reader, Object record) throws SchemaProjectorException {
+        if (writer.type() != reader.type()) {
+            throw new SchemaProjectorException("Schema type mismatch. Writer type: " + writer.type() + " and Reader type: " + reader.type());
+        }
+        Type writerValueType = writer.valueSchema().type();
+        Type readerValueType = reader.valueSchema().type();
+        if (writerValueType != readerValueType && !promote(writerValueType, readerValueType)) {
```

Comment:
And I think you want the same type of check here -- comparing the types is the first step, but you also need to check other parts of the schemas.

Alternatively, if you rely in the recursive call to `project(writer.valueSchema(), reader.valueSchema(), entry)` in the loop below, then wouldn't that already be doing the check for matching types such that you could get rid of this check?